src/stan/mcmc/hmc/base_hmc.hpp

```diff
@@ -127,7 +127,7 @@ class base_hmc : public base_mcmc {
         this->nom_epsilon_ = direction == 1 ? 2.0 * this->nom_epsilon_
                                             : 0.5 * this->nom_epsilon_;
 
-      if (this->nom_epsilon_ > 1e7)
+      if (!(this->nom_epsilon_ < 1e7))
```

Comment:
This inequality change is incorrect.  The current warning is meant to capture cases where the gradient vanishes (which occurs when the posterior density is flat, which will be encountered with probability > 0 only if the posterior is improper in at least some neighborhoods), _any_ stepwise will satisfy the initialization condition, and the heuristic initialization will result in an extremely large step size.

Reply:
I suppose it would make more sense to instead change the check that bails out at the start of the function
https://github.com/stan-dev/stan/blob/37ad8c36b83299d04fd9da824d324fa6517fc6e0/src/stan/mcmc/hmc/base_hmc.hpp#L84-L86

Reply:
The extreme step size occurs _during_ the initialization, hence the check immediately afterwards.